File: benchmarks/des-core/bench.ts
/**
 * Discrete-event simulation core benchmark — TypeScript / Node.
 *
 * Measures the throughput ceiling of the event loop that would sit at the heart
 * of the simulator: a priority queue over a transit-shaped event mix.
 *
 * Two representations are measured:
 *   "obj"   — idiomatic one-object-per-event
 *   "typed" — struct-of-arrays over TypedArrays
 *
 * Run directly (Node >= 22.18 strips types natively, no build step):
 *   node bench.ts obj   200000
 *   node bench.ts typed 1000000
 *
 * See ../README.md for results and interpretation.
 */

// ---------------------------------------------------------------- parameters

const NSTOPS = 4000;
const NVEH = 2000;
const VEH_STOPS = 40;
const LEGS = 3;

// Plain consts rather than `enum`: `enum` is not erasable syntax, so it would
// force a build step. Everything here stays runnable by `node bench.ts`.
const PAX_START = 0;
const PAX_ARRIVE = 1;
const PAX_BOARD = 2;
const PAX_ALIGHT = 3;
const VEH_STOP = 4;

type Kind = 0 | 1 | 2 | 3 | 4;

const M32 = 0xffffffff;

/** mulberry32. Bit-identical to the Python implementation in bench.py. */
function makeRng(seed: number): () => number {
  let a = seed >>> 0;
  return function (): number {
    a = (a + 0x6d2b79f5) | 0;
    let t = a;
    t = Math.imul(t ^ (t >>> 15), 1 | t);
    t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
    return (t ^ (t >>> 14)) >>> 0;
  };
}

export interface Result {
  runtime: string;
  mode: string;
  pax: number;
  events: number;
  seconds: number;
  events_per_sec: number;
  checksum: number;
  rss_mb?: number;
}

// -------------------------------------------------------- object-based core

interface Event {
  time: number;
  kind: Kind;
  ent: number;
  seq: number;
}

function runObjects(npax: number): { events: number; checksum: number } {
  const rng = makeRng(12345);
  const waiting = new Array<number>(NSTOPS).fill(0);
  const vehLoad = new Array<number>(NVEH).fill(0);
  const paxLeg = new Array<number>(npax).fill(0);
  const vehStopCount = new Array<number>(NVEH).fill(0);

  const heap: Event[] = [];
  let seq = 0;

  // Ties are broken by insertion sequence. A real simulator needs deterministic
  // tie-breaking, so its cost belongs in the measurement.
  const less = (a: Event, b: Event): boolean =>
    a.time < b.time || (a.time === b.time && a.seq < b.seq);

  function push(time: number, kind: Kind, ent: number): void {
    const node: Event = { time, kind, ent, seq: seq++ };
    heap.push(node);
    let i = heap.length - 1;
    while (i > 0) {
      const p = (i - 1) >> 1;
      if (!less(node, heap[p]!)) break;
      heap[i] = heap[p]!;
      i = p;
    }
    heap[i] = node;
  }

  function pop(): Event {
    const top = heap[0]!;
    const last = heap.pop()!;
    const n = heap.length;
    if (n > 0) {
      let i = 0;
      for (;;) {
        const l = 2 * i + 1;
        const r = l + 1;
        if (l >= n) break;
        let m = l;
        if (r < n && less(heap[r]!, heap[l]!)) m = r;
        if (!less(heap[m]!, last)) break;
        heap[i] = heap[m]!;
        i = m;
      }
      heap[i] = last;
    }
    return top;
  }

  for (let p = 0; p < npax; p++) push(rng() % 86400, PAX_START, p);
  for (let v = 0; v < NVEH; v++) push(rng() % 3600, VEH_STOP, v);

  let processed = 0;
  while (heap.length > 0) {
    const e = pop();
    processed++;
    const t = e.time;
    const ent = e.ent;
    switch (e.kind) {
      case PAX_START:
        waiting[rng() % NSTOPS]!++;
        push(t + 60 + (rng() % 600), PAX_ARRIVE, ent);
        break;
      case PAX_ARRIVE:
        waiting[rng() % NSTOPS]!--;
        push(t + 30 + (rng() % 900), PAX_BOARD, ent);
        break;
      case PAX_BOARD:
        vehLoad[rng() % NVEH]!++;
        push(t + 120 + (rng() % 1800), PAX_ALIGHT, ent);
        break;
      case PAX_ALIGHT:
        vehLoad[rng() % NVEH]!--;
        paxLeg[ent]!++;
        if (paxLeg[ent]! < LEGS) push(t + 60 + (rng() % 300), PAX_ARRIVE, ent);
        break;
      case VEH_STOP: {
        const s = rng() % NSTOPS;
        const dwell = 15 + (waiting[s]! > 0 ? 2 : 0) + (rng() % 45);
        vehStopCount[ent]!++;
        if (vehStopCount[ent]! < VEH_STOPS) {
          push(t + dwell + 60 + (rng() % 180), VEH_STOP, ent);
        }
        break;
      }
    }
  }

  return { events: processed, checksum: checksum(waiting, vehLoad, paxLeg, processed) };
}

// ----------------------------------------------- struct-of-arrays / typed core

function runTyped(npax: number): { events: number; checksum: number } {
  const rng = makeRng(12345);
  const waiting = new Int32Array(NSTOPS);
  const vehLoad = new Int32Array(NVEH);
  const paxLeg = new Uint8Array(npax);
  const vehStopCount = new Uint8Array(NVEH);

  // Peak queue depth is bounded: every passenger and vehicle holds at most one
  // scheduled event at a time.
  const cap = npax + NVEH + 16;
  const hTime = new Float64Array(cap);
  const hPay = new Int32Array(cap); // kind << 28 | ent
  const hSeq = new Uint32Array(cap);
  let n = 0;
  let seq = 0;

  function push(time: number, kind: Kind, ent: number): void {
    const pay = (kind << 28) | ent;
    const sq = seq++;
    let i = n++;
    while (i > 0) {
      const p = (i - 1) >> 1;
      if (hTime[p]! < time || (hTime[p]! === time && hSeq[p]! < sq)) break;
      hTime[i] = hTime[p]!;
      hPay[i] = hPay[p]!;
      hSeq[i] = hSeq[p]!;
      i = p;
    }
    hTime[i] = time;
    hPay[i] = pay;
    hSeq[i] = sq;
  }

  let popTime = 0;
  let popPay = 0;

  function pop(): void {
    popTime = hTime[0]!;
    popPay = hPay[0]!;
    n--;
    if (n > 0) {
      const lt = hTime[n]!;
      const lp = hPay[n]!;
      const ls = hSeq[n]!;
      let i = 0;
      for (;;) {
        const l = 2 * i + 1;
        const r = l + 1;
        if (l >= n) break;
        let m = l;
        if (r < n && (hTime[r]! < hTime[l]! || (hTime[r]! === hTime[l]! && hSeq[r]! < hSeq[l]!))) {
          m = r;
        }
        if (hTime[m]! > lt || (hTime[m]! === lt && hSeq[m]! > ls)) break;
        hTime[i] = hTime[m]!;
        hPay[i] = hPay[m]!;
        hSeq[i] = hSeq[m]!;
        i = m;
      }
      hTime[i] = lt;
      hPay[i] = lp;
      hSeq[i] = ls;
    }
  }

  for (let p = 0; p < npax; p++) push(rng() % 86400, PAX_START, p);
  for (let v = 0; v < NVEH; v++) push(rng() % 3600, VEH_STOP, v);

  let processed = 0;
  while (n > 0) {
    pop();
    processed++;
    const t = popTime;
    const kind = (popPay >>> 28) as Kind;
    const ent = popPay & 0x0fffffff;
    switch (kind) {
      case PAX_START:
        waiting[rng() % NSTOPS]++;
        push(t + 60 + (rng() % 600), PAX_ARRIVE, ent);
        break;
      case PAX_ARRIVE:
        waiting[rng() % NSTOPS]--;
        push(t + 30 + (rng() % 900), PAX_BOARD, ent);
        break;
      case PAX_BOARD:
        vehLoad[rng() % NVEH]++;
        push(t + 120 + (rng() % 1800), PAX_ALIGHT, ent);
        break;
      case PAX_ALIGHT:
        vehLoad[rng() % NVEH]--;
        paxLeg[ent]++;
        if (paxLeg[ent]! < LEGS) push(t + 60 + (rng() % 300), PAX_ARRIVE, ent);
        break;
      case VEH_STOP: {
        const s = rng() % NSTOPS;
        const dwell = 15 + (waiting[s]! > 0 ? 2 : 0) + (rng() % 45);
        vehStopCount[ent]++;
        if (vehStopCount[ent]! < VEH_STOPS) {
          push(t + dwell + 60 + (rng() % 180), VEH_STOP, ent);
        }
        break;
      }
    }
  }

  return { events: processed, checksum: checksum(waiting, vehLoad, paxLeg, processed) };
}

// --------------------------------------------------------------- equivalence

/**
 * Final-state fingerprint. All three implementations (obj, typed, Python) must
 * agree — that is what proves they performed identical work, rather than merely
 * similar-looking work.
 */
function checksum(
  waiting: ArrayLike<number>,
  vehLoad: ArrayLike<number>,
  paxLeg: ArrayLike<number>,
  processed: number,
): number {
  let a = 0;
  for (let i = 0; i < waiting.length; i++) a = (a + Math.imul(waiting[i]!, 31)) | 0;
  for (let i = 0; i < vehLoad.length; i++) a = (a + Math.imul(vehLoad[i]!, 17)) | 0;
  for (let i = 0; i < paxLeg.length; i++) a = (a + Math.imul(paxLeg[i]!, 7)) | 0;
  return ((a + processed) >>> 0) as number;
}

// ---------------------------------------------------------------------- main

const mode = process.argv[2] ?? "obj";
const npax = Number.parseInt(process.argv[3] ?? "200000", 10);
const fn = mode === "typed" ? runTyped : runObjects;

// Warm-up, so we measure steady-state JIT output rather than the interpreter tier.
fn(Math.min(20000, npax));

const t0 = process.hrtime.bigint();
const out = fn(npax);
const t1 = process.hrtime.bigint();
const secs = Number(t1 - t0) / 1e9;

const result: Result = {
  runtime: `node ${process.version}`,
  mode,
  pax: npax,
  events: out.events,
  seconds: Number(secs.toFixed(3)),
  events_per_sec: Math.round(out.events / secs),
  checksum: out.checksum,
  rss_mb: Number((process.memoryUsage().rss / 1048576).toFixed(1)),
};
console.log(JSON.stringify(result));

File: benchmarks/des-core/bench.py
"""Discrete-event simulation core benchmark -- Python.

Baseline counterpart to bench.ts. Same parameters, same event mix, same RNG
stream, same deterministic tie-breaking, so the final-state checksum must match
the TypeScript implementations exactly.

Uses heapq, whose C implementation favours Python -- this is deliberately the
strongest reasonable Python baseline, not a straw man.

Run:
    python bench.py 200000

See ../README.md for results and interpretation.
"""

import heapq
import json
import sys
import time

# ---------------------------------------------------------------- parameters

NSTOPS = 4000
NVEH = 2000
VEH_STOPS = 40
LEGS = 3

K_PAX_START, K_PAX_ARRIVE, K_PAX_BOARD, K_PAX_ALIGHT, K_VEH_STOP = 0, 1, 2, 3, 4

M32 = 0xFFFFFFFF


def make_rng(seed):
    """mulberry32. Bit-identical to the TypeScript implementation in bench.ts."""
    a = seed & M32

    def rng():
        nonlocal a
        a = (a + 0x6D2B79F5) & M32
        t = a
        t = ((t ^ (t >> 15)) * (1 | t)) & M32
        t = (((t + (((t ^ (t >> 7)) * (61 | t)) & M32)) & M32) ^ t) & M32
        return (t ^ (t >> 14)) & M32

    return rng


def checksum(waiting, veh_load, pax_leg, processed):
    """Final-state fingerprint; must equal the value produced by bench.ts."""
    a = 0
    for v in waiting:
        a = (a + v * 31) & M32
    for v in veh_load:
        a = (a + v * 17) & M32
    for v in pax_leg:
        a = (a + v * 7) & M32
    return (a + processed) & M32


def run(npax):
    rng = make_rng(12345)
    waiting = [0] * NSTOPS
    veh_load = [0] * NVEH
    pax_leg = [0] * npax
    veh_stop_count = [0] * NVEH

    heap = []
    push = heapq.heappush
    pop = heapq.heappop
    seq = 0

    # Ties are broken by insertion sequence, matching bench.ts.
    for p in range(npax):
        push(heap, (rng() % 86400, seq, K_PAX_START, p))
        seq += 1
    for v in range(NVEH):
        push(heap, (rng() % 3600, seq, K_VEH_STOP, v))
        seq += 1

    processed = 0
    while heap:
        t, _, kind, ent = pop(heap)
        processed += 1

        if kind == K_PAX_START:
            waiting[rng() % NSTOPS] += 1
            push(heap, (t + 60 + rng() % 600, seq, K_PAX_ARRIVE, ent))
            seq += 1
        elif kind == K_PAX_ARRIVE:
            waiting[rng() % NSTOPS] -= 1
            push(heap, (t + 30 + rng() % 900, seq, K_PAX_BOARD, ent))
            seq += 1
        elif kind == K_PAX_BOARD:
            veh_load[rng() % NVEH] += 1
            push(heap, (t + 120 + rng() % 1800, seq, K_PAX_ALIGHT, ent))
            seq += 1
        elif kind == K_PAX_ALIGHT:
            veh_load[rng() % NVEH] -= 1
            pax_leg[ent] += 1
            if pax_leg[ent] < LEGS:
                push(heap, (t + 60 + rng() % 300, seq, K_PAX_ARRIVE, ent))
                seq += 1
        else:
            s = rng() % NSTOPS
            dwell = 15 + (2 if waiting[s] > 0 else 0) + rng() % 45
            veh_stop_count[ent] += 1
            if veh_stop_count[ent] < VEH_STOPS:
                push(heap, (t + dwell + 60 + rng() % 180, seq, K_VEH_STOP, ent))
                seq += 1

    return processed, checksum(waiting, veh_load, pax_leg, processed)


def main():
    npax = int(sys.argv[1]) if len(sys.argv) > 1 else 200000

    run(min(20000, npax))  # warm-up, for parity with the TypeScript harness

    t0 = time.perf_counter()
    processed, chk = run(npax)
    secs = time.perf_counter() - t0

    print(json.dumps({
        "runtime": "python " + sys.version.split()[0],
        "mode": "heapq",
        "pax": npax,
        "events": processed,
        "seconds": round(secs, 3),
        "events_per_sec": round(processed / secs),
        "checksum": chk,
    }))


if __name__ == "__main__":
    main()
